src/opm/parser/eclipse/EclipseState/Schedule/Schedule.cpp

```diff
@@ -1428,16 +1428,48 @@ namespace {
       does not check or enforce in any way that this is done (i.e. it
       is not checked or verified that the well is initialized with any
       WCONxxxx keyword).
+
+      Update: See the discussion following the defintions of the SI factors, due
+        to a bad design we currently need the well to be specified with WCONPROD
+        / WCONHIST before WELTARG is applied, if not the units for the rates
+        will be wrong.
     */
 
     void Schedule::handleWELTARG( const SCHEDULESection& section ,
                                   const DeckKeyword& keyword,
                                   size_t currentStep,
                                   const ParseContext& parseContext, ErrorGuard& errors) {
         Opm::UnitSystem unitSystem = section.unitSystem();
-        double siFactorL = unitSystem.parse("LiquidSurfaceVolume/Time").getSIScaling();
-        double siFactorG = unitSystem.parse("GasSurfaceVolume/Time").getSIScaling();
-        double siFactorP = unitSystem.parse("Pressure").getSIScaling();
+        /*
+          double siFactorL = unitSystem.parse("LiquidSurfaceVolume/Time").getSIScaling();
+          double siFactorG = unitSystem.parse("GasSurfaceVolume/Time").getSIScaling();
+          double siFactorP = unitSystem.parse("Pressure").getSIScaling();
+        */
+
+        /*
+          Unit system handling in the UDA values has become a complete mess. The
+          point is that the UDA values can *optionally* carry a dimension object
+          with them, and then that unit system is transparaently used when
+          calling UDAValue::get<double>(). For UDA values which come from the
+          deck they are properly decorated with unitsystem, and things work as
+          they should. But when it comes to *explicitly setting* UDA values from
+          the API, as is done in the WELTARG implementation - things are quite
+          broken.
+
+          What currently "works" is:
+
+          - The well must be fully specified with WCONPROD / WCONHIST before the
+            WELTARG keyword is used.
+
+          - The unit conversion here in handleWELTARG is bypassed by setting the
+            conversion factors to 1.0, and the conversion which comes from the
+            dim objects internalized in the UDA objects will handle things on
+            return.
+        */
+
+        const double siFactorL = 1.0;
+        const double siFactorG = 1.0;
+        const double siFactorP = 1.0;
```

Comment:
With this, the unit from the WELTARG will be wrong.  For example, 
```
WELTARG
  PROD2 BHP 300
/
```

The well `PROD2` will have a BHP limit 300 Pascal instead of  300 Bar or 300 Psia, depending on the unit system. 

Currently, this change makes the model 2 fail because of the wrong BHP limits for an injector. 300 Pascal bhp will not be able to inject for that case.

Reply:
I am not following this issue so closely, please excuse me if I had some wrong comment.  Maybe with this change, you want the simulator to handle the unit or something else.

Reply:
My initial thought on this is that it should be ok. The limit of 300 in your example is stored in a `UDAValue`, which are actually supposed to be stored in deck units. Then the `WellInjectionProperties::controls()` method used the unit system to convert to SI after evaluation, and writes the result to the returned `InjectionControls` struct.

Have you checked that the limit actually ends up at 300 pascal in the above case?

Reply:
> Have you checked that the limit actually ends up at 300 pascal in the above case?

Compared with other pressure limits and bhp values, yes, it is 300 pascal in the simulator code.

Reply:
The real problem is not this change directly. For history mode, we use SI unit for the BHP limit, which causes the inconsistency here. Should we use `UDAValue` for the `WELTARG` target here?  

Anyway, I am testing to make the BHP limit for historical mode to be `UDAValue` first.

Reply:
Don't spend too much time on this - this was a temporary thing!

Reply:
It breaks model 2. See linuxbenchmarking.com ("Flow MPI Extra" broken since this was merged), so it must be addressed. If this is temporary only, can we make a temporary temporary fix...?

Reply:
It looks like with 
```
        const double siFactorL = unitSystem.parse("LiquidSurfaceVolume/Time").getSIScaling();
        const double siFactorP = unitSystem.parse("Pressure").getSIScaling();	
        const double siFactorG = unitSystem.parse("GasSurfaceVolume/Time").getSIScaling();
```

The running can be recovered.  Since it is not a lot of code, can we just recover the old way. When your final implementation is ready, then everything goes in together? 

How about this? @joakim-hove